app/src/main/java/org/openobservatory/ooniprobe/model/database/Url.java

```diff
@@ -56,6 +75,43 @@ public static Url checkExistingUrl(String input, String categoryCode, String cou
 		return url;
 	}
 
+	public static List<String> saveOrUpdate(List<Url> urls) {
+
+		Map<String, Url> resultUrlsMap = Maps.uniqueIndex(
+				urls,
+				input -> input.url
+		);
+
+		List<Url> existingUrls = Url.getExistingUrls(resultUrlsMap.keySet());
+
+		List<Url> existingUrlsToUpdate = Lists.newArrayList(Iterables.filter(existingUrls, input -> {
+			Url incomingChanges = resultUrlsMap.get(input.url);
+			if (incomingChanges != null) {
+				return ((!input.category_code.equals(incomingChanges.category_code)
+						&& !incomingChanges.category_code.equals("MISC"))
+						|| (!input.country_code.equals(incomingChanges.country_code)
+						&& !incomingChanges.country_code.equals("XX")));
+			} else {
+				return false;
+			}
```

Comment:
```suggestion
			if (incomingChanges == null) {
				return false;
			}
			return ((!input.category_code.equals(incomingChanges.category_code)
					&& !incomingChanges.category_code.equals("MISC"))
					|| (!input.country_code.equals(incomingChanges.country_code)
					&& !incomingChanges.country_code.equals("XX")));
```

We don't have a coding style manual that specifies that, however in the past both me and @hellais have expressed preference for not writing an `if .. else` chain where both arms return. The argument is that we'd like to have a single `if` whose arm expresses the less-interesting, often-but-not-in-this-case error related situation and the code falling through expressing the more interesting case.